crates/rust-analyzer/src/line_endings.rs

```diff
@@ -46,19 +46,7 @@ impl LineEndings {
         return (src, LineEndings::Dos);
 
         fn find_crlf(src: &[u8]) -> Option<usize> {
-            let mut search_idx = 0;
-            while let Some(idx) = find_cr(&src[search_idx..]) {
-                if src[search_idx..].get(idx + 1) != Some(&b'\n') {
-                    search_idx += idx + 1;
-                    continue;
-                }
-                return Some(search_idx + idx);
-            }
-            None
-        }
-
-        fn find_cr(src: &[u8]) -> Option<usize> {
-            src.iter().enumerate().find_map(|(idx, &b)| if b == b'\r' { Some(idx) } else { None })
+            src.iter().zip(src.iter().skip(1)).position(|it| it == (&b'\r', &b'\n'))
```

Comment:
I think the original idea here was that `find_cr` should auto-vectorise easily (call into memchr, really), but apparently that's not the case :-(

https://godbolt.org/z/JH8yaK

For max performance, we should pull `memchr` from crates.io here, but we don't need max performance now

bors r+

Reply:
https://godbolt.org/z/2j3urJ -- using a `&str` gives us `memchr`, as that is hard-coded in the stdlib.

Reply:
I suppose slice iter should override the default implementations of find/position and friends, but for some reason I (didn't check it) it most probably uses the default impl with try_fold() trickery under the hood which rustc is not powerful enough to minimize...
Btw, this function is heuristic since I suppose there might be wide unicode characters where the code point of `\r` is one of their byte footprint